### DISCUSSION TITLE
Alternative to `jax.core.new_main`

### DISCUSSION THREAD

**Discussion by nstarman** [jax-ml/jax#25372] [Q&A]
The library [`quax`](https://github.com/patrick-kidger/quax) relies ([code here](https://github.com/patrick-kidger/quax/blob/e62e5b30bd4bed81b0c27bee41272652ebfd6221/quax/_core.py#L310)) on the function `jax.core.new_main` , which was removed in v0.4.36 ([commit](https://github.com/jax-ml/jax/commit/c36e1f7c1ad4782060cbc8e8c596d85dfb83986f#diff-713d667c14b1dd075ec025dbd1cd8ea222e42c8b4c094ae05373d4d05b276dc5L1223)).
Does anyone know if there's an alternative means to substitute Jax tracers?

**Comment by mattjj** [comment]
Thanks for the question.

It's a bit tricky to explain in full. We plan to write an updated version of autodidax explaining the new core implementation, but we don't have that yet.

Maybe the best thing to do is to look at how that commit changed code involving new_main, e.g. [in ad.py](https://github.com/jax-ml/jax/commit/c36e1f7c1ad4782060cbc8e8c596d85dfb83986f#diff-af6216ea04ba29495025d310b4978b00ce0e6df6fcc579bb5bcdd9262132c9e7L79), since the changes needed on quax are likely analogous.

**Reply by nstarman** [reply]
Thank you! I'll take a look at that. See what I can figure out.

**Reply by mattjj** [reply]
This may just be me, but trying to run the quax tests I get a weird error:

```
mattjj at mattjj2 in ~/packages/quax on main
○ python tests/test_core.py                                                                                                                                                                         (base)
Traceback (most recent call last):
  File "/usr/local/google/home/mattjj/miniconda3/lib/python3.12/importlib/metadata/__init__.py", line 397, in from_name
    return next(cls.discover(name=name))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/google/home/mattjj/packages/quax/tests/test_core.py", line 11, in <module>
    import quax
  File "/usr/local/google/home/mattjj/packages/quax/quax/__init__.py", line 18, in <module>
    __version__ = importlib.metadata.version("quax")
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mattjj/miniconda3/lib/python3.12/importlib/metadata/__init__.py", line 889, in version
    return distribution(distribution_name).version
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mattjj/miniconda3/lib/python3.12/importlib/metadata/__init__.py", line 862, in distribution
    return Distribution.from_name(distribution_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mattjj/miniconda3/lib/python3.12/importlib/metadata/__init__.py", line 399, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: No package metadata was found for quax
```

Ever seen that one before?

**Reply by mattjj** [reply]
I guess I had to `pip install -e .`. (Thanks, Gemini!)

**Reply by mattjj** [reply]
This makes one test pass: https://github.com/patrick-kidger/quax/compare/main...mattjj:quax:stackless-fix

Namely `pytest tests/test_lora.py::test_complicated_dot`

Hopefully it starts to show the direction.

**Reply by nstarman** [reply]
Wow, thanks!

**Reply by mattjj** [reply]
Sorry, I meant the test was `pytest tests/test_lora.py::test_complicated_dot`! I think the others may have issues stemming from equinox (not sure yet).

**Reply by mattjj** [reply]
@patrick-kidger do you have a plan for updating the kidgerverse in light of c36e1f7c aka "stackless"? How can we be helpful? (Is there already a thread somewhere?)

**Reply by mattjj** [reply]
I updated https://github.com/patrick-kidger/quax/compare/main...mattjj:quax:stackless-fix to make all of tests/test_lora.py pass.

**Reply by mattjj** [reply]
Just pushed again, all of `pytest tests` pass now.

**Reply by mattjj** [reply]
Opened a draft PR: https://github.com/patrick-kidger/quax/pull/37

**Reply by patrick-kidger** [reply]
Oh awesome - thank you very much @mattjj !

I'm at neurips this week but I should be able to take a look afterwards.

I think I've updated most of the "kidgerverse" (😁) now, and Quax was the outstanding one as it hooked a lot more deeply into this functionality. So hopefully this will be it :)
